test/mpi/init/session_psets.c

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright (C) by Argonne National Laboratory
+ *     See COPYRIGHT in top-level directory
+ */
+
+#include "mpi.h"
+#include <stdio.h>
+#include "mpitest.h"
+
+/* This is adapted example 10.9 from MPI Standard 4.0 */
+
+static char use_pset_name[] = "SELF";
+
+int main(int argc, char *argv[])
+{
+    int errs = 0;
+    int i, n_psets, psetlen, rc, ret;
+    int valuelen;
+    int flag = 0;
+    char *pset_name = NULL;
+    char *info_val = NULL;
+    MPI_Session shandle = MPI_SESSION_NULL;
+    MPI_Info sinfo = MPI_INFO_NULL;
+    MPI_Group pgroup = MPI_GROUP_NULL;
+
+    rc = MPI_Session_init(MPI_INFO_NULL, MPI_ERRORS_RETURN, &shandle);
+    if (rc != MPI_SUCCESS) {
+        MTestPrintfMsg(1, "Could not initialize session, bailing out\n");
+        return -1;
+    }
+
+    MPI_Session_get_num_psets(shandle, MPI_INFO_NULL, &n_psets);
+    for (i = 0, pset_name = NULL; i < n_psets; i++) {
+        psetlen = 0;
+        MPI_Session_get_nth_pset(shandle, MPI_INFO_NULL, i, &psetlen, NULL);
+        if (psetlen <= 0) {
+            printf("MPI_Session_get_nth_pset: returns psetlen = %d\n", psetlen);
+            errs++;
+        }
+
+        pset_name = (char *) malloc(sizeof(char) * psetlen);
+        MPI_Session_get_nth_pset(shandle, MPI_INFO_NULL, i, &psetlen, pset_name);
+        if (strlen(pset_name) + 1 != psetlen) {
+            printf("MPI_Session_get_nth_pset: strlen(\"%s\") + 1 != %d\n", pset_name, psetlen);
+            errs++;
+        }
+
+        if (strstr(pset_name, use_pset_name) != NULL)
+            break;
```

Comment:
Why do we use `strstr` but not `strcmp`?

Reply:
Can we use a single preallocated `pset_name[MPI_MAX_PSET_NAME_LEN]` to avoid calling `malloc`/`free` multiple times?

Reply:
> 
> 
> Why do we use `strstr` but not `strcmp`?

From MPI standard text. User is likely to try the examples from standard, so I would like to keep this test very close to what is in the text.

Reply:
> 
> 
> Can we use a single preallocated `pset_name[MPI_MAX_PSET_NAME_LEN]` to avoid calling `malloc`/`free` multiple times?

We could. But I would like to keep it as original as we can.